src/containers/s21_list.h

```diff
@@ -0,0 +1,487 @@
+  list &operator=(const list &l) noexcept;
+  list &operator=(list &&l) noexcept;
+
+  void clear();
+  iterator insert(iterator pos, const_reference value);
+  void push_back(const_reference data) noexcept;
+  void pop_back();
+  void push_front(const_reference value) noexcept;
+  void pop_front();
+  void merge(list &other);
+  void splice(const_iterator pos, list &other);
+  void sort();
+  void reverse() noexcept;
+  void unique();
+  void erase(iterator pos);
+  const_reference front();
+  const_reference back();
+
+ private:
+  Node *m_head;
+  Node *m_tail;
+  size_type m_size;
+  Node *m_end;
+};
+}  // namespace s21
+
+template <typename T>
+s21::list<T>::ListIterator::ListIterator() noexcept : current_(nullptr) {}
+
+template <typename T>
+s21::list<T>::ListIterator::ListIterator(Node *node) noexcept
+    : current_(node) {}
+
+template <typename T>
+typename s21::list<T>::reference s21::list<T>::ListIterator::operator*() const {
+  return current_->value_;
+};
+
+template <typename T>
+typename s21::list<T>::ListIterator &s21::list<T>::ListIterator::operator++() {
+  current_ = current_->next_;
+  return *this;
+}
+
+template <typename T>
+typename s21::list<T>::ListIterator s21::list<T>::ListIterator::operator++(
+    int) {
+  ListIterator temp(current_);
+  current_ = current_->next_;
+  return temp;
+}
+
+template <typename T>
+typename s21::list<T>::ListIterator &s21::list<T>::ListIterator::operator--() {
+  current_ = current_->prev_;
+  return *this;
+}
+
+template <typename T>
+typename s21::list<T>::ListIterator s21::list<T>::ListIterator::operator--(
+    int) {
+  ListIterator temp(current_);
+  current_ = current_->prev_;
+  return temp;
+}
+
+template <typename T>
+bool s21::list<T>::ListIterator::operator==(
+    const s21::list<T>::ListIterator &other) const {
+  return current_ == other.current_;
+}
+
+template <typename T>
+bool s21::list<T>::ListIterator::operator!=(const ListIterator &other) const {
+  return !(*this == other);
+}
+
+template <typename T>
+typename s21::list<T>::const_reference
+s21::list<T>::ListConstIterator::operator*() const {
+  return this->current_->value_;
+}
+
+template <typename T>
+typename s21::list<T>::iterator s21::list<T>::begin() noexcept {
+  return iterator(m_head ? this->m_head : this->m_end);
+}
+
+template <typename T>
+typename s21::list<T>::const_iterator s21::list<T>::cbegin() const noexcept {
+  return const_iterator(m_head ? this->m_head : this->m_end);
+}
+
+template <class T>
+typename s21::list<T>::iterator s21::list<T>::end() noexcept {
+  return iterator(m_end);
+}
+
+template <typename T>
+typename s21::list<T>::const_iterator s21::list<T>::cend() const noexcept {
+  return const_iterator(m_end);
+}
+
+template <typename T>
+s21::list<T>::list() noexcept : m_head(nullptr), m_tail(nullptr), m_size(0) {
+  m_end = new Node(T());
+  update_end();
+}
+
+template <typename T>
+s21::list<T>::list(size_type n) : m_head(nullptr), m_tail(nullptr), m_size(n) {
```

Comment:
Нейминг - почему бы "размер" не назвать "размером?

```suggestion
s21::list<T>::list(size_type size) : m_head(nullptr), m_tail(nullptr), m_size(size) {
```